solver/src/oneinch_solver/api.rs

```diff
@@ -0,0 +1,367 @@
+//! 1Inch HTTP API client implementation.
+//!
+//! For more information on the HTTP API, consult:
+//! <https://docs.1inch.io/api/quote-swap>
+//! <https://api.1inch.exchange/swagger/ethereum/>
+
+use anyhow::{ensure, Result};
+use ethcontract::{H160, U256};
+use reqwest::{
+    header::{HeaderValue, ACCEPT},
+    Client, ClientBuilder, IntoUrl, Url,
+};
+use serde::{
+    de::{Deserializer, Error as _},
+    Deserialize,
+};
+use std::{
+    borrow::Cow,
+    fmt::{self, Display, Formatter},
+    time::Duration,
+};
+
+/// A slippage amount.
+#[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
+pub struct Slippage(f64);
+
+impl Slippage {
+    /// Creates a slippage amount from the specified percentage.
+    pub fn percentage(amount: f64) -> Result<Self> {
+        // 1Inch API only accepts a slippage from 0 to 50.
+        ensure!(
+            (0. ..=50.).contains(&amount),
+            "slippage outside of [0%, 50%] range"
+        );
+        Ok(Slippage(amount))
+    }
+
+    /// Creates a slippage amount from the specified basis points.
+    pub fn basis_points(bps: u16) -> Result<Self> {
+        let percent = (bps as f64) / 100.;
+        Slippage::percentage(percent)
+    }
+}
+
+impl Display for Slippage {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// A 1Inch API quote query parameters.
+///
+/// These parameters are currently incomplete, and missing parameters can be
+/// added incrementally as needed.
+pub struct SwapQuery {
```

Comment:
Is it possible to model buy orders?

Reply:
Good question...

Reply:
According to https://docs.1inch.io/api/quote-swap I don't see any obvious way of doing it :disappointed: